### PR TITLE
adding ipfs / arweave support

### LINKED ISSUE #1711: Add ipfs and arweave storage type support
Support all 3 storage types : url, ipfs, arweave

Screenshot 2022-09-27 164004

For ipfs

Screenshot 2022-09-27 163336

Structure : 
```
 {
      "type": "ipfs",
      "hash": "XXXX"
    }
``` 

For arweave 

Screenshot 2022-09-27 163848

Structure:
```
{
  "type": "arweave",
  "transactionId": "a4qJoQZa1poIv5guEzkfgZYSAD0uYm7Vw4zm_tCswVQ"
}
```

The validation is done the same way on the `fileInfo` endpoint, using the relevant structure.

WIP: 

- [X] add new fields to publish form
- [X] handle `storageType` selection UI
- [x] integrate `Ipfs` and `Arweave` type from `oceanjs` lib (opened an issue for ipfs, [missing interface](https://github.com/oceanprotocol/ocean.js/pull/1640))
- [x] handle txs with different `storageType` in publish
- [x] change edit (UI & Logic) accordantly

### PR DESCRIPTION
**Pull request by EnzoVezzaro**
- Fixes #1711 .

Changes proposed in this PR:

- created storage type dropdown (Url / IPFS / Arweave)
- added storage type selection on publish
- added storage type selection on edit
- updated `@oceanprotocol/lib@2.5.2`

WIP: 

- [X] UI redesign fileInput
- [x] fix IPFS typing [issue](https://github.com/oceanprotocol/ocean.js/pull/1640)
- [x] wait for fix provider: [IPFS](https://github.com/oceanprotocol/provider/issues/558) and [Arwave](https://github.com/oceanprotocol/provider/issues/555)
- [x] validation and edit form
- [x] waiting for https://github.com/oceanprotocol/provider/issues/555 to test `ARWEAVE`
- [x] fix failing test after merged main
- [x] finish suggestions on review
  - [x] redesign UI uploader
  - [x] error messages and catch errors
  - [x] cleanup and functionality test
  - [x] tests

TO TEST:

- IPFS: you can upload a file in IPFS using [web3.storage](https://web3.storage) (to get CID)
- Arweave: you can use this[ little app](https://hv4gxzchk24cqfezebn3ujjz6oy2kbtztv5vghn6kpbkjc3vg4rq.arweave.net/PXhr5EdWuCgUmSBbuiU587GlBnmde1MdvlPCpIt1NyM/) to upload a file into Arweave (to get the transactionID)

## Publish

Before validation:
Screenshot 2022-11-11 at 10 44 47

After validation:
Screenshot 2022-11-14 at 11 16 57

## Edit
When entered in edit form:
Screenshot 2022-11-14 at 10 50 13

After clicking on "x":
Screenshot 2022-11-14 at 10 50 19

### CODE REVIEW

File: src/@utils/provider.ts [hunk 1]
```diff
@@ -85,16 +87,44 @@ export async function getFileDidInfo(
 
 export async function getFileUrlInfo(
```
Comment (kremalicious) [inline src/@utils/provider.ts:88]: same [as in that other PR](https://github.com/oceanprotocol/market/pull/1732#discussion_r995408419), we are extending this method to check for more than a URL, hence "Url" should be removed from the name.

File: src/@utils/provider.ts [hunk 2]
```diff
@@ -85,16 +87,44 @@ export async function getFileDidInfo(
...
   try {
-    const fileUrl: UrlFile = {
-      type: 'url',
-      index: 0,
-      url,
-      method: 'get'
+    let response
+    switch (storageType) {
+      case 'ipfs':
+        // eslint-disable-next-line no-case-declarations
```
Comment (kremalicious) [inline src/@utils/provider.ts:97]: why this overwrite? Just call the variable `fileIpfs`

File: src/@utils/provider.ts [hunk 3]
```diff
@@ -85,16 +87,44 @@ export async function getFileDidInfo(
...
+          type: 'ipfs',
+          hash: url
+        }
+
+        response = await ProviderInstance.getFileInfo(fileIPFS, providerUrl)
+        console.log(response, fileIPFS)
+
+        break
+      case 'arweave':
+        // eslint-disable-next-line no-case-declarations
```
Comment (kremalicious) [inline src/@utils/provider.ts:108]: why is this here?

File: src/@utils/provider.ts [hunk 4]
```diff
@@ -85,16 +87,44 @@ export async function getFileDidInfo(
 
 export async function getFileUrlInfo(
   url: string,
```
Comment (kremalicious) [inline src/@utils/provider.ts:89]: same here, we add more than URL checking in here so parameter should not be named `url`.

File: src/@utils/provider.ts [hunk 5]
```diff
@@ -85,16 +87,44 @@ export async function getFileDidInfo(
...
-  providerUrl: string
-): Promise<FileInfo[]> {
+  providerUrl: string,
+  storageType: string
+): Promise<any[]> {
```
Comment (kremalicious) [inline src/@utils/provider.ts:92]: please more precise than `any`

File: src/@utils/provider.ts [hunk 6]
```diff
@@ -85,16 +87,44 @@ export async function getFileDidInfo(
...
+        // eslint-disable-next-line no-case-declarations
+        const fileArweave: Arweave = {
+          type: 'arweave',
+          transactionId: url
+        }
+
+        response = await ProviderInstance.getFileInfo(fileArweave, providerUrl)
+        break
+      default:
+        // eslint-disable-next-line no-case-declarations
```
Comment (kremalicious) [inline src/@utils/provider.ts:117]: same as above

File: src/@utils/url.ts [hunk 7]
```diff
@@ -1,5 +1,11 @@
+import * as isIPFS from 'is-ipfs'
+
 export function sanitizeUrl(url: string) {
   const u = decodeURI(url).trim().toLowerCase()
   const isAllowedUrlScheme = u.startsWith('http://') || u.startsWith('https://')
   return isAllowedUrlScheme ? url : 'about:blank'
 }
+
+export function isCID(value: string) {
```
Comment (kremalicious) [inline src/@utils/url.ts:9]: 1. this seems overengineered, as it is pretty simple calling `isIPFS.cid(value)`
2. IPFS does not deal with URLs, but CIDs. Hence this method should not be in `@utils/url`, or the file need to be renamed

File: src/@utils/provider.ts (hunk 4 above)
Comment (kremalicious) [reply src/@utils/provider.ts:89]: this has not been solved. Again, `url` is now only one way of storage. So we can't name everything `url`. Use e.g. `file` here and use `url` naming only for the actual thing, URLs.

File: src/components/@shared/FormInput/InputElement/ContainerInput/index.tsx [hunk 8]
```diff
@@ -75,6 +75,9 @@ export default function ContainerInput(props: InputProps): ReactElement {
           name={`${field.name}[0].url`}
           checkUrl={false}
           isLoading={isLoading}
+          storageType={
+            'url' /* TODO: allow different storageType. https://github.com/oceanprotocol/market/issues/1711 */
```
Comment (kremalicious) [inline src/components/@shared/FormInput/InputElement/ContainerInput/index.tsx:79]: what is this? I think we should actually use this component instead of a dropdown, and then make sure we can use a lot of items beside each other. Would be way better UI than this dropdown

**Comment by kremalicious** [comment]
Some UI feedback: the labeling here is confusing, all of them describe "Files":

Screenshot 2022-11-09 at 09 11 34

Screenshot 2022-11-09 at 09 11 55

So would suggest for the label, "File URL", "File CID", "File Transaction ID"

Sentence like "Please make sure that the endpoint is accessible over the internet and is not protected by a firewall or by credentials." does not make sense for IPFS or Arweave so please remove.

For IPFS & possibly arweave we also need to be more precise in what is allowed, and most likely detect certain things and inform publisher. Like, are folder CIDs allowed?

Also while we are at it to make these help texts shorter, the sentence "For a compute dataset, your file should match the file type required by the algorithm, and should not exceed 1 GB in file size." should only appear for compute assets.

**Comment by kremalicious** [comment]
what are your test assets? Tried to publish IPFS file on Mumbai but my dataset does not get indexed:

https://market-git-feature-issue-1711-add-ipfs-arw-755c8b-oceanprotocol.vercel.app/asset/did:op:18aaddcb061487dfa41450253587c8f368532b25aaa65ada2169337ae034b1cf

**Comment by kremalicious** [comment]
as a quick UI wireframe, this is how the visual hierarchy should be done, the whole thing describes the "File":

Screenshot 2022-11-09 at 10 03 24

**Comment by EnzoVezzaro** [comment]
> as a quick UI wireframe, this is how the visual hierarchy should be done to make it clear what belongs to each other, the whole thing describes the "File", and not storage type + file as separate things.
> 
> Screenshot 2022-11-09 at 10 03 24
> 
> Where then after successful validation, we show our existing "file added" UI, with addition of showing which file storage was used:
> 
> Screenshot 2022-11-09 at 10 53 27

Got it 👍🏽 great review, thanks. I'll start working on it 💪🏽

> Like, are folder CIDs allowed? Which version we support, v0 or v1, or both? Cause in the placeholder example for CID you show the v0 format.

Mmm..... 🤔 I'll test all this cases to see how to handle them, I'll try IPNS also even though I guess there's no problem there. Do you know if Arweave have advance any folder capabilities? I know Ardrive kind of uses something similar to IPFS folders. 

> Sentence like "Please make sure that the endpoint is accessible over the internet and is not protected by a firewall or by credentials." does not make sense for IPFS or Arweave so please remove for those cases.
> Also while we are at it to make these help texts shorter, the sentence "For a compute dataset, your file should match the file type required by the algorithm, and should not exceed 1 GB in file size." should only appear for compute assets.

Ok 👍🏽

**Comment by alexcos20** [comment]
only ipfs is supported, no ipns support yet.
If you think it's needed, let's have a chat

**Comment by jamiehewitt15** [comment]
First test, things looked good but then after publishing the asset I got this error. Looks like an odd one, I'll try publishing again

image

**Comment by EnzoVezzaro** [comment]
Mmmm... weird 🤔 just tried both IPFS / Aweave without issues when publishing. The only think that I found was the same issue we currently have on Priced Assets. Not sure if it's related. Do you know anything else about this issue @bogdanfazakas ?

IPFS free, ok:
Screenshot 2022-11-23 at 07 48 07

Arweave (ignore title, I used an Arweave file here) priced, not ok:
Screenshot 2022-11-23 at 07 48 04

**Comment by jamiehewitt15** [comment]
I ran another test published on IPFS and it worked fine this time. I'll let you know if I'm able to reproduce it.

**Comment by jamiehewitt15** [comment]
One thing we should do is explain to people how to go about publishing with Arweave or IPFS. I've created an [issue](https://github.com/oceanprotocol/docs/issues/1158) for this in the docs. When that's done we should add a link to the docs section in the tooltip.

**Comment by jamiehewitt15** [comment]
Also tested publishing a Arweave asset and it worked well 👍

It would be great if we could add a test to this to keep the overall test coverage going up.

**Comment by EnzoVezzaro** [comment]
> It would be great if we could add a test to this to keep the overall test coverage going up.

We already have [a test ](https://github.com/oceanprotocol/market/pull/1765/files#diff-f6e79a26f5f9ed776f8e8e6cfb51979153d882dd39a65b5a1762b76bad4f596a) for both IPFS and Arwaeve. I'm kind of rushing a bit this because I need to keep going with the integration of graphql and on-chain data which depends on this PR. Maybe we can add more tests afterwards.

**Comment by jamiehewitt15** [comment]
> We already have [a test ](https://github.com/oceanprotocol/market/pull/1765/files#diff-f6e79a26f5f9ed776f8e8e6cfb51979153d882dd39a65b5a1762b76bad4f596a)for both IPFS and Arwaeve. I'm kind of rushing a bit this because I need to keep going with the integration of graphql and on-chain data which depends on this PR. Maybe we can add more tests afterwards.

Yeah, fair enough. I was just seeing that the overall test coverage is going down.

**Comment by jamiehewitt15** [comment]
Why is it only possible to use a URL for the sample file? People may also want to use IPFS/ Arweave for that

**Comment by EnzoVezzaro** [comment]
> Why is it only possible to use a URL for the sample file? People may also want to use IPFS/ Arweave for that

Maybe it'll be implemented in the future, but for this task is [out of scope](https://github.com/oceanprotocol/market/issues/1711#issuecomment-1293534127).

**Comment by jamiehewitt15** [comment]
> Maybe it'll be implemented in the future, but for this task is [out of scope](https://github.com/oceanprotocol/market/issues/1711#issuecomment-1293534127).

Ok sure, seems fair enough

File: src/components/Publish/_utils.ts [hunk 9]
```diff
@@ -139,18 +139,25 @@ export async function transformPublishFormToDdo(
...
+        [files[0].type === 'ipfs'
+          ? 'hash'
+          : files[0].type === 'arweave'
+          ? 'transactionId'
+          : 'url']: files[0].url,
         method: 'GET'
       }
     ]
   }
+  console.log(file)
```
Comment (bogdanfazakas) [inline src/components/Publish/_utils.ts:159]: missed this 😄

**Review by jamiehewitt15 — APPROVED** [review]
Have done quite a few tests now and can't find any faults. Works well, good job

File: src/components/Publish/_validation.ts [hunk 10]
```diff
@@ -51,16 +85,34 @@ const validationService = {
...
-            'GoogleNotSupported',
-            'Google Drive is not a supported hosting service. Please use an alternative.',
-            (value) => {
-              return !value?.toString().includes('drive.google')
-            }
-          ),
-        // TODO: require valid file only when URL is given
+        url: Yup.string().test((value, context) => {
+          const { type } = context.parent
+          let validField
```
Comment (bogdanfazakas) [inline src/components/Publish/_validation.ts:90]: wouldn't it be useful to have a helper to validate, and use it for both publish and edit, we can reduce code duplication like so

**Review by bogdanfazakas — APPROVED** [review]
In terms of functionality all good from my tests, good job with the PR @EnzoVezzaro 👍

File: src/components/Publish/_utils.ts (hunk 9 above)
Comment (EnzoVezzaro) [reply src/components/Publish/_utils.ts:159]: 😅 thanks

File: src/components/Publish/_validation.ts (hunk 10 above)
Comment (EnzoVezzaro) [reply src/components/Publish/_validation.ts]: done. Can you please check one last time, to make sure it's all fine 👍🏽 thanks bud

**Review by bogdanfazakas — APPROVED** [review]

File: src/components/Asset/Edit/FormEditMetadata.tsx [hunk 11]
```diff
@@ -83,23 +86,48 @@ export default function FormEditMetadata({
 
   return (
     <Form>
-      {data.map(
```
Comment (jamiehewitt15) [inline src/components/Asset/Edit/FormEditMetadata.tsx:86]: @EnzoVezzaro How come you decided to remove the mapping?

Comment (EnzoVezzaro) [reply src/components/Asset/Edit/FormEditMetadata.tsx:86]: just to have better control over them like we do in publish. For example [here](https://github.com/oceanprotocol/market/blob/feature/issue-1724-graphql-onchain-storage-support/src/components/Publish/Services/index.tsx#L0-L1), in publish every fields is added manually in case we need to do some special logic on a specific field. Also not sure if the mapping breaks the [graphql / on-chain integration](https://github.com/oceanprotocol/market/pull/1794). Sorry If I didn't included the `paymentCollector` field, I lost it.
